src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -196,129 +239,156 @@ case class WorkflowActor(workflow: WorkflowDescriptor, backend: Backend)
+      val updatedData = data.addPersisting(callKey)
+      transition(WorkflowFailed) using updatedData
+    case Event(message @ CallAborted(callKey), data) if callKey.isPersistedRunning(data) =>
       // Something funky's going on if aborts are coming through while the workflow's still running. But don't second-guess
       // by transitioning the whole workflow - the message is either still in the queue or this command was maybe
       // cancelled by some external system.
-      Await.result(persistStatus(callKey, ExecutionStatus.Aborted), Duration.Inf)
+      persistStatusThenAck(callKey, ExecutionStatus.Aborted, sender(), message)
       logger.warn(s"Call ${callKey.scope.name} was aborted but the workflow should still be running.")
-      stay()
+      val updatedData = data.addPersisting(callKey)
+      stay() using updatedData
   }
 
-  when(WorkflowFailed) {
-    case Event(GetFailureMessage, msg: FailureMessage) =>
-      sender() ! msg
-      stay()
+  when(WorkflowSucceeded) { FSM.NullFunction }
+
+  when(WorkflowFailed) { FSM.NullFunction }
+
+  when(WorkflowAborted) { FSM.NullFunction }
+
+  private def callCompletedWhileAborting(callKey: OutputKey, outputs: CallOutputs, returnCode: Int): State = {
+    awaitCallComplete(callKey, outputs, returnCode)
+    val nextState = if (isWorkflowAborted) transition(WorkflowAborted) else stay()
+    val updatedData = stateData.addPersisting(callKey)
+    nextState using updatedData
   }
 
-  when(WorkflowSucceeded) {
-    case Event(Terminate, _) =>
-      logger.debug(s"WorkflowActor is done, shutting down.")
-      context.stop(self)
-      stay()
+  onTransition {
+    case WorkflowSubmitted -> WorkflowRunning =>
+      stateData.startMode.get.start(this)
   }
 
   when(WorkflowAborting) {
-    case Event(AbortComplete(callKey), NoFailureMessage) =>
-      Await.result(persistStatus(callKey, ExecutionStatus.Aborted, None), Duration.Inf)
-      if (isWorkflowAborted) goto(WorkflowAborted) using NoFailureMessage else stay()
-    case Event(CallFailed(callKey, returnCode, failure), NoFailureMessage) =>
-      Await.result(persistStatus(callKey, ExecutionStatus.Failed, returnCode), Duration.Inf)
-      if (isWorkflowAborted) goto(WorkflowAborted) using NoFailureMessage else stay()
-    case Event(CallCompleted(callKey, outputs, returnCode), NoFailureMessage) =>
-      awaitCallComplete(callKey, outputs, returnCode)
-      if (isWorkflowAborted) goto(WorkflowAborted) using NoFailureMessage else stay()
-    case m =>
+    case Event(CallCompleted(callKey, outputs, returnCode), data) if callKey.isPersistedRunning(data) =>
+      callCompletedWhileAborting(callKey, outputs, returnCode)
+    case Event(CallCompleted(collectorKey: CollectorKey, outputs, returnCode), data) if !collectorKey.isPending(data) =>
+      // Collector keys don't have to be in Running state.
+      val updatedData = data.addPersisting(collectorKey)
```

Comment:
`val updatedData = data.addPersisting(collectorKey)` can just be `data.addPersisting(collectorKey)`.